app/project_type/git.py

```diff
@@ -144,45 +136,63 @@ def _fetch_project(self):
-        self._git_remote_command_executor.execute(fetch_command, cwd=self._repo_directory)
+        self._execute_git_command_in_repo_and_raise_on_failure(
+            'fetch --update-head-ok {} {}'.format(self._remote, self._branch),
+            'Could not fetch specified branch "{}" from remote "{}".'.format(self._branch, self._remote)
+        )
 
         # Validate and convert the user-specified hash/refspec to a full git hash
-        self._hash = self._execute_in_repo_and_raise_on_failure(
-            'git rev-parse {}'.format(self._hash),
+        self._hash = self._execute_git_command_in_repo_and_raise_on_failure(
+            'rev-parse {}'.format(self._hash),
             'Could not rev-parse "{}" to a commit hash.'.format(self._hash)
         ).strip()
 
         # Save this hash as a local ref. Named local refs are necessary for slaves to fetch correctly from the master.
         # The local ref will be passed on to slaves instead of the user-specified branch.
         self._local_ref = 'refs/clusterrunner/' + self._hash
-        update_ref_command = 'git update-ref {} {}'.format(self._local_ref, self._hash)
-        self._execute_in_repo_and_raise_on_failure(update_ref_command, 'Could not update local ref.')
+        update_ref_command = 'update-ref {} {}'.format(self._local_ref, self._hash)
+        self._execute_git_command_in_repo_and_raise_on_failure(update_ref_command, 'Could not update local ref.')
 
         # The '--' option acts as a delimiter to differentiate values that can be "tree-ish" or a "path"
-        reset_command = 'git reset --hard {} --'.format(self._hash)
-        self._execute_in_repo_and_raise_on_failure(reset_command, 'Could not reset Git repo.')
+        reset_command = 'reset --hard {} --'.format(self._hash)
+        self._execute_git_command_in_repo_and_raise_on_failure(reset_command, 'Could not reset Git repo.')
 
-        self._execute_in_repo_and_raise_on_failure('git clean -dfx', 'Could not clean Git repo.')
+        self._execute_git_command_in_repo_and_raise_on_failure('clean -dfx', 'Could not clean Git repo.')
 
-    def _execute_in_repo_and_raise_on_failure(self, command, message):
+    def _execute_git_command_in_repo_and_raise_on_failure(self, command, message='Error executing git command.'):
         """
+        :param command: The git command to execute, e.g., "fetch origin"
+        :type command: string
+        :param message: The human readable error message to log if the command fails
+        :type message: string
+        :return: The output of the process (stdout and stderr)
         :rtype: string
         """
-        return self._execute_and_raise_on_failure(command, message, self._repo_directory)
+        strict_host_key_setting = 'yes' if Configuration['git_strict_host_key_checking'] else 'no'
+        git_ssh_args = '-o BatchMode=yes -o StrictHostKeyChecking={}'.format(strict_host_key_setting)
```

Comment:
Just googled it, and you're right about the BatchMode=yes and PasswordAuthentication=no.

The UserKnownHostsFile, however, is not quite the same. If you don't set the UserKnownHostsFile=/dev/null, but have StrictHostKeyChecking=no, there is a (very) small corner case you can run into where a host that already exists in ~/.ssh/known_hosts has changed its host key, and would consequently warn you about a man-in-the-middle attack.